java/src/main/java/com/google/protobuf/ZeroCopyEncoder.java

```diff
@@ -0,0 +1,647 @@
+      flush();
+
+      // Write a read-only view of this buffer. The new buffer has its own position and limit.
+      handler.onDataEncoded(value, offset, length);
+      return;
+    }
+
+    while(offset < length) {
+      int capacity = capacity();
+      if (capacity == 0) {
+        // The buffer is full, flush it and try again.
+        flush();
+        continue;
+      }
+
+      // Determine the number of bytes that can be written before another flush is required.
+      int remaining = length - offset;
+      int bytesToWrite = Math.min(remaining, capacity);
+
+      getOrCreateBuffer().put(value, offset, bytesToWrite);
+      offset += bytesToWrite;
+    }
+  }
+
+  @Override
+  public void writeRawBytes(ByteString value, int offset, int length) throws IOException {
+    if (length == 0) {
+      return;
+    }
+
+    if (value instanceof LiteralByteString) {
+      LiteralByteString lbs = (LiteralByteString) value;
+      writeRawBytes(lbs.bytes, lbs.getOffsetIntoBytes(), lbs.size());
+    } else if (value instanceof UnsafeByteString) {
+      writeRawBytes(value.asReadOnlyByteBuffer());
+    } else {
+      for(ByteBuffer buf : value.asReadOnlyByteBufferList()) {
+        writeRawBytes(buf);
+      }
+    }
+  }
+
+  @Override
+  public void writeTag(int fieldNumber, int wireType) throws IOException {
+    writeRawVarint32(WireFormat.makeTag(fieldNumber, wireType));
+  }
+
+  @Override
+  public void writeRawVarint32(int value) throws IOException {
+    while (true) {
+      if ((value & ~0x7F) == 0) {
+        writeRawByte(value);
+        return;
+      } else {
+        writeRawByte((value & 0x7F) | 0x80);
+        value >>>= 7;
+      }
+    }
+  }
+
+  @Override
+  public void writeRawVarint64(long value) throws IOException {
+    while (true) {
+      if ((value & ~0x7FL) == 0) {
+        writeRawByte((int) value);
+        return;
+      } else {
+        writeRawByte(((int) value & 0x7F) | 0x80);
+        value >>>= 7;
+      }
+    }
+  }
+
+  @Override
+  public void writeRawLittleEndian32(int value) throws IOException {
+    if(capacity() < INTEGER_BYTES) {
+      flush();
+    }
+    getOrCreateBuffer();
+    buffer.order(ByteOrder.LITTLE_ENDIAN);
+    buffer.putInt(value);
+    buffer.order(ByteOrder.LITTLE_ENDIAN);
+  }
+
+  @Override
+  public void writeRawLittleEndian64(long value) throws IOException {
+    if(capacity() < LONG_BYTES) {
+      flush();
+    }
+    getOrCreateBuffer();
+    buffer.order(ByteOrder.LITTLE_ENDIAN);
+    buffer.putLong(value);
+    buffer.order(ByteOrder.LITTLE_ENDIAN);
+  }
+
+  private int position() {
+    return buffer == null ? 0 : buffer.position();
+  }
+
+  private int capacity() {
+    return bufferSize - position();
+  }
+
+  private ByteBuffer getOrCreateBuffer() {
+    if (buffer == null) {
+      // Guarantee that it has a backing array.
+      buffer = ByteBuffer.wrap(new byte[bufferSize]);
+      buffer.order(ByteOrder.BIG_ENDIAN);
```

Comment:
is this necessary? why not just always LITTLE if everything else is written as bytes